gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java

```diff
@@ -157,32 +164,112 @@ public void handleResponse(WriteObjectResponse response) {
   }
 
   @Override
-  public void startUpload(InputStream pipeSource) {
-    // Given that the two ends of the pipe must operate asynchronous relative
-    // to each other, we need to start the upload operation on a separate thread.
+  public void startUpload(InputStream pipeSource) throws IOException {
+    return;
+  }
+
+  @Override
+  public boolean isOpen() {
+    return (writeBuffer != null);
+  }
+
+  @Override
+  public synchronized int write(ByteBuffer buffer) throws IOException {
```

Comment:
I found this method hard to understand. Can you please add more comments and try to simplify this method